### PR TITLE
Update resource_builtin_role_assignment resource to ignore assignments defined in the server side

### LINKED ISSUE #362: BuiltInRole deploy not idempotent
**Terraform Version**
Terraform v0.14.4
Terraform Grafana Provider 1.12.0
Grafana Enterprise v8.3.5

**Affected Resource(s)**
grafana_builtin_role_assignment

**Terraform config**
```
      "grafana_builtin_role_assignment": {
         "editor": {
            "builtin_role": "Editor",
            "roles": [
               {
                  "global": false,
                  "uid": "${grafana_role.report_viewer.uid}"
               }
            ]
         }
      },
      "grafana_role": {
         "report_viewer": {
            "description": "Can view reports",
            "name": "Report Viewer",
            "permissions": [
               {
                  "action": "reports:read",
                  "scope": "reports:*"
               }
            ],
            "version": 1
         }
      }
```

**Expected Behavior**
`terraform deploy` should create the role + assignment to Editor on first deploy, and make no changes on second deploy.

**Actual Behavior**
As expected on first deploy, but on second deploy all built-in-role-assignments for Editor are destroyed, including the default assignments, in our case the assignment of datasources:explore to Editor, and only the built-in role assignment specified in our config is recreated.

### PR DESCRIPTION
**Pull request by vtorosyan**
## Context

When Grafana starts, by default it creates built-in role assignments for fixed roles. Additionally, users can create built-in role assignments via API. In these scenarios, when the resource is not properly imported, terraform will try to destroy what is in server side, but not in the configuration.

The change ensures that this does not happen as it can cause unintended behaviour for users.

## Implementation details

The only workaround, with the given resources is to do in advance check to see if the fetched assignments has been ever in the state, or is present in the configuration, otherwise ignore it.

## Testing

I have manually tested the following scenarios:
- Having built-in role assignments in server, and creating a new assignment in terraform
- Updating the above created assignments after it's first time creation
- Removing roles from the above assignments
- Destroying entire resource -> in this scenario what has not been defined in the config, won't be destroyed from the backend.

Additionally, the newly added test covers the scenario mentioned in the issue.

## Related issues

Fixes https://github.com/grafana/terraform-provider-grafana/issues/362

### CODE REVIEW

File: grafana/resource_builtin_role_assignment_test.go [hunk 1]
```diff
@@ -48,6 +48,147 @@ func TestAccBuiltInRoleAssignment(t *testing.T) {
...
 
+func TestAccBuiltInRoleAssignmentUpdate(t *testing.T) {
+	CheckEnterpriseTestsEnabled(t)
+
+	var br gapi.BuiltInRoleAssignment
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: testAccProviderFactories,
+		CheckDestroy:      testAccBuiltInRoleAssignmentCheckDestroy(&br),
```
Comment (julienduchesne) [inline grafana/resource_builtin_role_assignment_test.go]: Should the `CheckDestroy` function also check that the default assignements weren't destroyed?

**Review by julienduchesne — APPROVED** [review]
The solution seems like a good one to me and it's well tested. I can't test it on my side but it LGTM!

File: grafana/resource_builtin_role_assignment_test.go [hunk 2]
```diff
@@ -48,6 +48,147 @@ func TestAccBuiltInRoleAssignment(t *testing.T) {
...
+	if err != nil {
+		return fmt.Errorf("error creating role: %s", err)
+	}
+	_, err = client.NewBuiltInRoleAssignment(gapi.BuiltInRoleAssignment{
+		BuiltinRole: "Viewer",
+		RoleUID:     role1.UID,
+		Global:      false,
+	})
+	if err != nil {
+		return fmt.Errorf("error creating built-in role assigntment: %s", err)
```
Comment (Jguer) [inline grafana/resource_builtin_role_assignment_test.go]:
```suggestion
		return fmt.Errorf("error creating built-in role assigntment: %w", err)
```

nit: just so error un/wrapping still succeeds

File: grafana/resource_builtin_role_assignment_test.go [hunk 3]
```diff
@@ -48,6 +48,147 @@ func TestAccBuiltInRoleAssignment(t *testing.T) {
...
+	}
+	_, err = client.NewBuiltInRoleAssignment(gapi.BuiltInRoleAssignment{
+		BuiltinRole: "Viewer",
+		RoleUID:     role2.UID,
+		Global:      false,
+	})
+	if err != nil {
+		return fmt.Errorf("error creating built-in role assigntment: %s", err)
```
Comment (Jguer) [inline grafana/resource_builtin_role_assignment_test.go]:
```suggestion
		return fmt.Errorf("error creating built-in role assigntment: %w", err)
```

nit: just so error un/wrapping still succeeds

File: grafana/resource_builtin_role_assignment_test.go [hunk 4]
```diff
@@ -48,6 +48,147 @@ func TestAccBuiltInRoleAssignment(t *testing.T) {
...
+			},
+		},
+	}
+	role1, err := client.NewRole(r1)
+	if err != nil {
+		return fmt.Errorf("error creating role: %s", err)
+	}
+	role2, err := client.NewRole(r2)
+	if err != nil {
+		return fmt.Errorf("error creating role: %s", err)
```
Comment (Jguer) [inline grafana/resource_builtin_role_assignment_test.go]:
```suggestion
		return fmt.Errorf("error creating role: %w", err)
```

nit: just so error un/wrapping still succeeds

File: grafana/resource_builtin_role_assignment_test.go [hunk 5]
```diff
@@ -48,6 +48,147 @@ func TestAccBuiltInRoleAssignment(t *testing.T) {
...
+		Global:  true,
+		Permissions: []gapi.Permission{
+			{
+				Action: "users:create",
+			},
+		},
+	}
+	role1, err := client.NewRole(r1)
+	if err != nil {
+		return fmt.Errorf("error creating role: %s", err)
```
Comment (Jguer) [inline grafana/resource_builtin_role_assignment_test.go]:
```suggestion
		return fmt.Errorf("error creating role: %w", err)
```

nit: just so error un/wrapping still succeeds

**Review by Jguer — APPROVED** [review]

File: grafana/resource_builtin_role_assignment_test.go (hunk 1 above)
Comment (vtorosyan) [reply grafana/resource_builtin_role_assignment_test.go]: Really good catch. I actually just realized that `testAccBuiltInRoleAssignmentCheckDestroy` checks nothing, as the pointer to `&br` never updates to the created built-in role assignment, so we always pass an empty struct to the method, which returns success on it's turn as there is indeed no assignment for empty built-in role. 

Working on a fix.

Comment (vtorosyan) [reply grafana/resource_builtin_role_assignment_test.go]: 5566c8b4691e45c25d4b1383e155229156eda19f

File: grafana/resource_builtin_role_assignment.go [hunk 6]
```diff
@@ -102,8 +102,20 @@ func ReadBuiltInRole(ctx context.Context, d *schema.ResourceData, meta interface
...
 	}
 
+	stateRoles, configRoles, err := collectRoles(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
 	roles := make([]interface{}, 0)
 	for _, br := range brRole {
+		// It is possible on the server side there are roles assigned to the built-in role which were never in Terraform state,
+		// and are not now in the current configuration. To prevent unintended behaviour, such as destroying remove resources,
```
Comment (inkel) [inline grafana/resource_builtin_role_assignment.go]: I'm not sure I understand what _destroying remove resources_ means, could you rephrase this? Thank you.

**Review by inkel — COMMENTED** [review]
I've just left a minor request. My initial review looks good, but I want to give it another round to fully understand what are we testing here.

Thank you so much for your hard work!

File: grafana/resource_builtin_role_assignment.go (hunk 6 above)
Comment (vtorosyan) [reply grafana/resource_builtin_role_assignment.go]: Good point, I can see it was unclear and wrong. 396491caf72db7241154901dd00ce33d83069d9d is this a bit better?

Comment (inkel) [reply grafana/resource_builtin_role_assignment.go]: Yup, much better, thank you!
